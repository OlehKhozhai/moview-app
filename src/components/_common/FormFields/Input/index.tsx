import React from 'react';
import { useField } from 'formik';
import cn from 'classnames';

import ErrorMessage from 'components/_common/ErrorMessage';
import styles from './styles.module.scss';

type InputProps = {
  label: string;
  name: string;
  className?: string;
} & React.InputHTMLAttributes<HTMLInputElement>;

const Input: React.FC<InputProps> = ({ label, type, ...props }) => {
  const [showPassword, setShowPassword] = React.useState(false);
  const [field, meta] = useField({ ...props });
  const isPasswordType = type === 'password';

  React.useEffect(() => {
    if (isPasswordType) {
      setShowPassword(isPasswordType);
    }
  }, [isPasswordType]);

  const handleShowPassword = () => {
    setShowPassword((prevState) => !prevState);
  };

  return (
    <div className={styles.root}>
      <label className={styles.label}>{label}</label>
      <input
        className={cn(styles.input, { [styles.errorInput]: meta.touched && meta.error })}
        {...field}
        {...props}
        type={isPasswordType && showPassword ? 'password' : type}
      />
      {isPasswordType && (
        <span className={styles.passwordVisibilityIcon} onClick={handleShowPassword}>
          {showPassword ? '^' : '?'}
        </span>
      )}
      {meta.touched && meta.error && (
        <ErrorMessage text={meta.error} className={styles.errorMessage} />
      )}
    </div>
  );
};

export default React.memo(Input);
