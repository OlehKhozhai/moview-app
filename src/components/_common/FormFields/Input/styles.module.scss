@import 'assets/styles/mixins/colors';

.root {
  display: flex;
  flex-flow: column;
  padding-bottom: 30px;
  position: relative;
}

.label {
  text-transform: uppercase;
  color: $secondary-color;
}

.input {
  margin-top: 15px;
  padding: 15px;
  border-radius: 4px;
  font-size: 22px;
  border: 1px solid transparent;
  background-color: $primary-light-bg;
  color: $primary-color;

  &:focus {
    outline: none;
  }
}

input[type='date']::-webkit-calendar-picker-indicator {
  opacity: 0.6;
  filter: invert(0.8);
  cursor: pointer;
}

.errorInput {
  border: 1px solid $primary-border;
}

.errorMessage {
  bottom: 10px;
}

.passwordVisibilityIcon {
  position: absolute;
  right: 0;
  top: 45%;
}
