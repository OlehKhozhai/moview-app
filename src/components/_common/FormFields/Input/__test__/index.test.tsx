import React from 'react';
import formik from 'formik';
import { create } from 'react-test-renderer';

import Input from '../index';

jest.mock('formik', () => ({
  ...jest.requireActual('formik'),
}));

describe('Input component', () => {
  test('should render Input component by default', () => {
    jest.spyOn(formik, 'useField').mockReturnValue([
      { name: 'name', value: 'value', onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: 'value', touched: true },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);

    const component = create(<Input name="name" label="label" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Input component with type of password', () => {
    jest.spyOn(formik, 'useField').mockReturnValue([
      { name: 'name', value: 'value', onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: 'value', touched: true },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);

    const component = create(<Input name="name" label="label" type="password" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Input component with error message', () => {
    jest.spyOn(formik, 'useField').mockReturnValue([
      { name: 'name', value: 'value', onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: 'value', touched: true, error: 'Error message' },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);

    const component = create(<Input name="name" label="label" type="password" />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
