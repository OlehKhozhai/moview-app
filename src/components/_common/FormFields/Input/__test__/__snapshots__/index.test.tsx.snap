// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render Input component by default 1`] = `
<div
  className="root"
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input"
    name="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    value="value"
  />
</div>
`;

exports[`Input component should render Input component with error message 1`] = `
<div
  className="root"
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input errorInput"
    name="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    type="password"
    value="value"
  />
  <span
    className="passwordVisibilityIcon"
    onClick={[Function]}
  >
    ^
  </span>
  <p
    className="root errorMessage"
  >
    Error message
  </p>
</div>
`;

exports[`Input component should render Input component with type of password 1`] = `
<div
  className="root"
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input"
    name="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    type="password"
    value="value"
  />
  <span
    className="passwordVisibilityIcon"
    onClick={[Function]}
  >
    ?
  </span>
</div>
`;
